### PR TITLE
Prepare for 64 PMP config registers

### PR DESCRIPTION
**Pull request by alistair23**
### Pull Request Overview

This PR was supposed to add support for 64 PMP config register on RISC-V now that it is supported by the spec.

Instead this patch has turned more info converting the riscv-csr library to use the new inline assembly macro and to allow dynamic array access to the RISC-V registers.

This PR does a few things:
 - Converts the PMP CSRs to an array (to allow simplifying the PMP code)
 - Uses the new inline assembly. Unfortunately Rust won't let use access CSR dynamically like we used to (see https://github.com/rust-lang/rust/issues/73220) so we now have a giant if statement to see which one to access. I couldn't see any other way to make the value const.
 - We also implement the register functions on specific types (u32) instead of generics. This is because with generics we can't dynamically access arrays of CSRs with the inline assembly.
 - Finally we clean up some of the PMP code.

### Testing Strategy

None yet.

### TODO or Help Wanted

Comments please, this is uglier then I would like it to be.

### Documentation Updated

- [X] Updated the relevant files in `/docs`, or no updates are required.

### Formatting

- [X] Ran `make prepush`.

### CODE REVIEW

File: arch/rv32i/src/pmp.rs [hunk 1]
```diff
@@ -531,115 +408,115 @@ impl kernel::mpu::MPU for PMPConfig {
...
                                         + csr::pmpconfig::pmpcfg::x0::CLEAR
                                         + csr::pmpconfig::pmpcfg::a0::TOR,
                                 );
-                                csr::CSR.pmpaddr4.set((start as u32) >> 2);
+                                csr::CSR.pmpaddr[4].set((start as u32) >> 2);
 
                                 // Set access to end address
-                                csr::CSR.pmpcfg1.set(cfg_val << 8 | csr::CSR.pmpcfg0.get());
-                                csr::CSR.pmpaddr5.set((start as u32 + size as u32) >> 2);
+                                csr::CSR.pmpcfg[1].set(cfg_val << 8 | csr::CSR.pmpcfg[0].get());
```
Comment (a-pronin) [inline arch/rv32i/src/pmp.rs:448]: should be pmpcfg[1]
please see all places where changes are needed in https://github.com/tock/tock/pull/1932

File: libraries/riscv-csr/src/csr.rs [hunk 2]
```diff
@@ -33,73 +69,272 @@ impl<T: IntLike, R: RegisterLongName> ReadWriteRiscvCsr<T, R> {
...
+    pub fn matches_any(&self, field: FieldValue<u32, R>) -> bool {
+        self.get() & field.mask != u32::zero()
     }
 
     #[inline]
-    pub fn matches_all(&self, field: FieldValue<T, R>) -> bool {
+    pub fn matches_all(&self, field: FieldValue<u32, R>) -> bool {
         self.get() & field.mask == field.value
     }
 }
```
Comment (gendx) [inline libraries/riscv-csr/src/csr.rs:340]: All of these functions look very similar to what's already available in `libraries/tock-register-interface/src/registers.rs`. Is there any reason not to use that?

File: arch/rv32i/src/csr/mod.rs [hunk 3]
```diff
@@ -23,26 +23,8 @@ pub struct CSR {
...
-    pub pmpaddr11: ReadWriteRiscvCsr<u32, pmpaddr::pmpaddr::Register>,
-    pub pmpaddr12: ReadWriteRiscvCsr<u32, pmpaddr::pmpaddr::Register>,
-    pub pmpaddr13: ReadWriteRiscvCsr<u32, pmpaddr::pmpaddr::Register>,
-    pub pmpaddr14: ReadWriteRiscvCsr<u32, pmpaddr::pmpaddr::Register>,
-    pub pmpaddr15: ReadWriteRiscvCsr<u32, pmpaddr::pmpaddr::Register>,
+    pub pmpcfg: [ReadWriteRiscvCsr<u32, pmpconfig::pmpcfg::Register>; 4],
+    pub pmpaddr: [ReadWriteRiscvCsr<u32, pmpaddr::pmpaddr::Register>; 16],
     pub mie: ReadWriteRiscvCsr<u32, mie::mie::Register>,
     pub mscratch: ReadWriteRiscvCsr<u32, mscratch::mscratch::Register>,
     pub mepc: ReadWriteRiscvCsr<u32, mepc::mepc::Register>,
```
Comment (gendx) [inline arch/rv32i/src/csr/mod.rs:30]: It looks like this definition could use the `register_structs` macro defined in `libraries/tock-register-interface/src/macros.rs`. Is there any reason not to use it?

File: libraries/riscv-csr/src/csr.rs (hunk 2 above)
Comment (bradjc) [reply libraries/riscv-csr/src/csr.rs:340]: Reading a CSR requires an assembly instruction, not accessing a memory location.

File: arch/rv32i/src/csr/mod.rs (hunk 3 above)
Comment (bradjc) [reply arch/rv32i/src/csr/mod.rs:30]: Alistair is just updating what is already there, I think it is fine to not add additional changes.

Comment (gendx) [reply arch/rv32i/src/csr/mod.rs:30]: Note: my question was beyond the scope of this PR, to see if there's any improvement that could be made in `libraries/tock-register-interface` to improve adoption of this library, not only in `chips/` but in `arch/` as well.

File: libraries/riscv-csr/src/csr.rs (hunk 2 above)
Comment (gendx) [reply libraries/riscv-csr/src/csr.rs:340]: A lot of the logic still overlaps with methods such as `FieldValue::matches_all`. The only difference that I see is in the implementation of the `set` and `get` methods - all of the other methods are derived from it in the same manner as in the `libraries/tock-register-interface` library.

It looks like the `ReadAble` and `WriteAble` traits that I sketched in https://github.com/tock/tock/pull/1766#issue-405159759 could be beneficial here as well.

**Comment by ppannuto** [comment]
bors d=alistair23,gendx

**Pull request merged**